### PR TITLE
[stable/concourse] Upgrade to Concourse 4

### LINKED ISSUE #6873: [stable/concourse] Upgrade chart to support Concourse 4.0
Concourse 4.0 has been released with some breaking changes, requiring changes to the chart. I'd like to start working on this, but don't want to duplicate work in case someone else is already working on it.

@frodenas @william-tran Is this safe to work on? Should the chart support both 3.x and 4.0?

Breaking changes:
- [x] Migrate `CONCOURSE_BASIC_AUTH_*` to `CONCOURSE_ADD_LOCAL_USER`, `CONCOURSE_MAIN_TEAM_LOCAL_USER`, and `CONCOURSE_MAIN_TEAM_ALLOW_ALL_USERS` (don't forget to change the NOTES.txt to match the new basic auth model)
- [x] Migrate `CONCOURSE_GITHUB_AUTH_*` to `CONCOURSE_GITHUB_*` and `CONCOURSE_MAIN_TEAM_GITHUB_*`
- [x] Migrate `CONCOURSE_GITLAB_AUTH_*` to `CONCOURSE_GITLAB_*` and `CONCOURSE_MAIN_TEAM_GITLAB_*`
- [x] Migrate `CONCOURSE_GENERIC_OAUTH_*` to `CONCOURSE_OAUTH_*` and `CONCOURSE_MAIN_TEAM_OAUTH_*`

New features:
- [ ] `CONCOURSE_RESOURCE_TYPE_CHECKING_INTERVAL`
- [ ] `CONCOURSE_VAULT_CACHE`

New authentication modes:
- [x] CloudFoundry
- [x] LDAP

Pre-existing authentication modes not already in chart:
- [x] OIDC

### PR DESCRIPTION
**Pull request by wendorf**
**What this PR does / why we need it**:

Supports the recently-released Concourse 4, and drops support for versions earlier than 4.0.0

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #6873

**Special notes for your reviewer**:

This is a big PR, potentially too big. I've split up the pieces of work into individual commits, so can easily remove pieces if needed. I've tried to capture all the desires in #6873, including removing old parameters not supported in 4.0. I have left one deprecated parameter, `CONCOURSE_POSTGRES_DATA_SOURCE`, because it's a little painful to change all of the documentation and usage to use the new split-up parameters, and this was already becoming a big PR. If that's something we would like, though, I can certainly do the work and update this PR.

### CODE REVIEW

**Comment by rubenv** [comment]
Haven't been able to test the PR itself, but from just looking at the diff it looks like this will fix all the pain I've had from trying to use the current chart for Concourse 4. Big step forward and the way to go.

Thanks @wendorf :+1:

**Comment by wendorf** [comment]
Rebased on top of the latest changes to the Concourse chart.

**Comment by william-tran** [comment]
@wendorf sorry been away on vacation, thanks for all the work you've put into this

**Comment by arbourd** [comment]
I was using this this week with defaults and haven't run into any issues :)

Thank you so much @wendorf

File: stable/concourse/templates/web-deployment.yaml
Comment (regisbelson) [inline stable/concourse/templates/web-deployment.yaml]: There's an extra `-` in front of `value` here, it breaks the chart when you use Datadog

**Review by regisbelson — CHANGES_REQUESTED** [review]

**Comment by kurlzor** [comment]
I was able to upgrade our production instance after fixing that typo for the Datadog env var. Thanks a lot for your work !

File: stable/concourse/templates/web-deployment.yaml
Comment (wendorf) [reply stable/concourse/templates/web-deployment.yaml]: Thanks for the catch! Fixed.

File: stable/concourse/README.md
Comment (romankartgh) [inline stable/concourse/README.md:242]: Would it be better to avoid adding this parameter all together? Turns out OpenID provider in coreos/dex doesn't support group claims: https://github.com/coreos/dex. We integrated with Azure, but there is to way to retrieve the list of groups user belongs to. So this parameter doesn't seem to have any effect on the behavior.

File: stable/concourse/values.yaml
Comment (romankartgh) [inline stable/concourse/values.yaml:917]: Same comment as above

File: stable/concourse/README.md
Comment (wendorf) [reply stable/concourse/README.md:242]: I haven't actually used OIDC Authentication at all, so I'm not sure—I've just folded in all of the flags Concourse supports. I recommend asking the Concourse team directly, probably at https://discuss.concourse-ci.org/ .

Comment (romankartgh) [reply stable/concourse/README.md:242]: I filed https://github.com/concourse/concourse/issues/2538, let's see what the response is

Comment (romankartgh) [reply stable/concourse/README.md:242]: sounds like people got this working with okta but not azure. this seems to be okay.
you can disregard my original comment, dan. thanks!

**Review by romankartgh — APPROVED** [review]
we tested this on our side. migrated all parameters from 3.14.1 to 4.0.0. introduced new auth parameters - they get propagated fine for 4.0.0.

File: stable/concourse/templates/web-deployment.yaml
Comment (william-tran) [inline stable/concourse/templates/web-deployment.yaml]: Could we get rid of these cmd line args and use the env var `CONCOURSE_AWS_REGION` instead? From the output of `concourse web --help` it looks like that env var is used in both AWS SSM and Secretsmanager cases. Other than that we're looking good!

**Review by william-tran — CHANGES_REQUESTED** [review]

File: stable/concourse/templates/web-deployment.yaml
Comment (wendorf) [reply stable/concourse/templates/web-deployment.yaml]: According to the binary I have, the variable it wants to use is `AWS_REGION` (not `CONCOURSE_AWS_REGION`, which would be a little nicer), which seemed like a "heavy" env var to set if you want to configure either the `--aws-secretsmanager-region` or `--aws-ssm-region`. It would mean that configuring one implicitly (and perhaps surprisingly) configures the other one. This is "expected behavior" from Concourse's perspective, but it felt weird to support it, which is why I left it.

If my argument isn't convincing, just let me know and I'll change it :)

**Comment by pms1969** [comment]
It may or not be related to this change, but I've been unable to get this working on a fresh install without manual intervention.  When the workers start up, they try to retire the worker which returns a non-zero exit code.  The startup loops, until it eventually fails.  Don't know if this is a change from the non 4.0.0 binaries since this is the first time I've tried to use the chart, and wanted to move directly to 4.0.0

File: stable/concourse/templates/web-deployment.yaml
Comment (william-tran) [reply stable/concourse/templates/web-deployment.yaml]: @wendorf you're right it's `AWS_REGION`, and more importantly, it should be avoided due to this unfortunate issue: https://github.com/concourse/concourse/issues/2191. Lets stick with the cmd line args, but would it be possible to pass the value directly into the args rather than using an env var to propagate it?

Comment (wendorf) [reply stable/concourse/templates/web-deployment.yaml]: Sounds good. Just pushed the change.

**Comment by wendorf** [comment]
I just bumped the PR for the recently-released Concourse 4.1:

* Bumped the app and image versions from 4.0.0 -> 4.1.0
* Updated the `secrets.localUserAuth.localUsers` to match the new description in 4.1 (more detail around bcrypting)
* Did not add the new `CONCOURSE_OIDC_HOSTED_DOMAINS` env var since, using the current structure, it would require putting an array in a Secret and I didn't have the motivation to figure that out :)
* I diffed the `concourse web --help` output between the two versions, and I don't believe there are any other important changes (there are some changed parameters that already weren't part of the chart).

**Comment by wendorf** [comment]
@pms1969 Without more specific detail, it's hard to help, but since this chart is mostly just funneling parameters to Concourse I'd lean toward assuming it's a problem between your specific config and Concourse itself. You can probably get good help at https://discuss.concourse-ci.org/ or https://github.com/concourse/concourse/issues, but if you discover a problem with this PR, I'd love to know what we can fix!

**Comment by william-tran** [comment]
@wendorf something's up with how concourse-web is trying to connect to the DB, take a look at the CI log: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/helm_charts/6951/pull-charts-e2e/13157/

```
I0831 20:14:36.457] failed to perform migrations: creating migration table: dial tcp 127.0.0.1:5432: connect: connection refused
```

**Comment by wendorf** [comment]
Yeah, if I switch to setting `CONCOURSE_POSTGRES_HOST`, `CONCOURSE_POSTGRES_USER`, `CONCOURSE_POSTGRES_PASSWORD`, and `CONCOURSE_POSTGRES_DATABASE` it works. I'd submit that fix quickly, but I want to understand if `CONCOURSE_POSTGRES_DATA_SOURCE` is totally broken, especially since we suggest people do that when providing an external data source.

**Comment by wendorf** [comment]
Looks like the `CONCOURSE_POSTGRES_DATA_SOURCE` env var (and its flag) are not correctly used in 4.1. I've written up more details in concourse/concourse#2549

I think the choices here are:

1) hold off on updating to 4.1, revert to 4.0, and merge the PR sooner
2) switch the chart to use the specific (e.g. `CONCOURSE_POSTGRES_HOST`) env vars instead of `CONCOURSE_POSTGRES_DATA_SOURCE`. This would require also changing the interface for the external postgres source, since a single secret key is no longer sufficient. `CONCOURSE_POSTGRES_DATA_SOURCE` is officially deprecated anyway, so this work will have to be done at some point.
3) wait to hear back on the issue and see if either a) I'm crazy and got this wrong or b) the Concourse team ships a point release to fix the bug. This would potentially delay merging this PR for a while.

I'm not really tied to any specific choice here. I think I'd prefer option 2, but it is a bit more work for me. @william-tran what are your thoughts?

**Comment by wendorf** [comment]
In the mean time, I've done the revert so this PR is working. The open question now is to merge as-is or spend time to make it work for 4.1.

**Comment by pms1969** [comment]
@wendorf Yeah, didn't think it had anything to do with your changes, was just posting in case you, or someone else had seen it and knew what might be wrong.  Your efforts here are greatly appreciated.

**Comment by william-tran** [comment]
@wendorf given `CONCOURSE_POSTGRES_DATA_SOURCE` is officially deprecated, even though it may be a bug that it doesn't work in 4.1, we're going to have to change it sooner or later. Looking at the release notes for 4.1 as well, there are a few fixes to issues that I wouldn't want to be without:

> Fixed an annoying issue affecting deployments with multiple web nodes. We were configuring an in-memory store for the Dex auth flow, meaning your login session would only work on one ATC at a time, and could fail partway through the redirect dance.

> The concourse/concourse Docker image now contains the file command, which is useful for...working. Without it the btrfs volume driver setup would fail. Sorry about that, and thanks to @ElfoLiNk for submitting the PR!

> We've fixed a scary container snowballing failure mode that could happen when check containers failed to create. This was a somewhat hairy bug; see concourse #2454 for more information.

If you have the cycles to make the change this week, go for it. If not, LMK and I can make a commit on my own fork, and you can cherry pick that into this PR.

**Comment by wendorf** [comment]
> [I think it's time to just remove the flag as it's been deprecated since v2.7.2 (!). (Sorry for breaking it though.)](https://github.com/concourse/concourse/issues/2549#issuecomment-418417800)

I'll just make the changes (hopefully today).

**Comment by wendorf** [comment]
Just pushed up a new commit (`Breaking: Invididual DB env vars`) that removes the use of CONCOURSE_POSTGRES_DATA_SOURCE. I've tested installing with both a pre-provided database instance as well as using the postgres chart to create a database for me, and both seem to work just fine.

If there are no issues with the key structure I've chosen, I think this is ready to merge.

**Comment by william-tran** [comment]
@wendorf I notice now that many non-secret values that relate to features that have secret values, have now moved into the secrets object. This makes it harder to `kubectl describe` something and see the configuration that it's running with, minus the secret stuff.  This also makes using secrets outside of helm more difficult as there's lots more values that you need to deal with; rather than using your values.yaml for those. What was your motivation for this change?

**Comment by wendorf** [comment]
The main motivation was wanting to keep all related configuration together, e.g. all oidc config in one block. Because they were already nested under secrets, I felt compelled to make everything a secret.

I'd personally prefer to remove the secrets YAML namespace altogether and just put things in the Secret as necessary:

```
oidcAuth:
  enabled:
  displayName:
  issuer:
  clientId:
  clientSecret:
  scope:
  groupsKey:
  caCert:
  skipSslValidation:
```

I like the simplicity, but this does have the downside of not making it clear what keys are stored as secrets. To keep the secrets section as only secrets, the structure needs to be split up:

```
oidcAuth:
  enabled:
  displayName:
  issuer:
  clientId:
  scope:
  groupsKey:
  caCert:
  skipSslValidation:

secrets:
  oidcAuth:
    clientSecret:
```

This gets confusing (for me at least) because to properly configure OIDC (or anything else that's partially secret) I need to look in very different places (potentially hundreds of lines of YAML).

Another option would be to call out secrets explicitly:

```
oidcAuth:
  enabled:
  displayName:
  issuer:
  clientId:
  scope:
  groupsKey:
  caCert:
  skipSslValidation:
  secrets:
    clientSecret:
```

This is a decent middle-ground, and certainly something I'd be okay with.

A final concern is that, especially with the new auth methods, I'm not an expert on what parameters should be considered secret. I _think_ oidcAuth.groupsKey is not secret, but I'm not positive.

What are your thoughts? I'll defer to what you think is most appropriate.

**Comment by wendorf** [comment]
Another consideration: The PR as is doesn't have presence guards around environment variables, so it sets a lot of environment variables to empty strings. Generally this isn't a problem, but does cause some weird scenarios like:

```
→ fly -t od teams -d
name         users                                                                        groups
main         cf:,github:,gitlab:,ldap:,oauth:,oidc:,local:mypassword  cf:,cf:,github:,github:my-org,gitlab:,ldap:,oauth:,oidc:
```

It's "cleaner"/"more correct" to add a bunch of presence guards like:

```
{{ if .cfUsers }}
- name: CONCOURSE_MAIN_TEAM_CF_USER
  value: {{ join "," .cfUsers }}
{{ end }}
```

but it makes the template much more verbose. @william-tran Do you have opinions here?

**Comment by chrishiestand** [comment]
@wendorf While I don't have any specific examples of where lack of guards is a problem, it does seem to move the problem of validating values downstream to the consumer of these variables. I could imagine annoying-to-debug problems occurring if empty values are consumed. It seems to me that it would be a good idea to add presence guards to the helm chart, despite the ugliness.

**Comment by william-tran** [comment]
On the presence guards, a template helper could be devised, thought needing to use `dict` to pass in params is also a bit ugly in its own way. Maybe `list`? 

On separating secrets, I'm putting a lot of weight on the use case of not having any secrets managed by helm values at all. See https://github.com/helm/charts/issues/3208#issuecomment-355419260 and https://github.com/helm/charts/pull/3254 for context. This is an important use case for improving the security posture of your CI system, which has the keys to everything. Given that use case of managing secrets outside of helm values, the ideal scenario is where secrets is kept to a minimum; they only contain credentials, so that the management overhead of these values is kept to a minimum, and so taking this option for security's sake doesn't become that much more difficult. 

I can also see that this forces the chart maintainers to think about what actually is a secret and what isn't, but this is an important thing for us to make that determination for the benefit of chart users.

In terms of moving from `secrets.oidcAuth.clientsecret` and `oidcAuth.secrets.clientsecret`, this is at best "6 of 1, half a dozen of the other", which doesn't warrant the change in structure.

**Comment by william-tran** [comment]
That said, I've only heard two voices on the refactoring of secrets, mine and yours @wendorf. @nexeck  thoughts? Anyone else?

**Comment by AndrewMorsillo** [comment]
I installed this in my cluster and it seems to work great except that the web UI continually redirects to 127.0.0.1 instead of the proper IP/domain. Did I miss a setting @wendorf or is that a bug in the chart?

**Comment by antonu17** [comment]
Looks like values web.env are entirely ignored in deployment template

**Comment by wendorf** [comment]
@antonu17 `web.env` was renamed to `web.extraEnv` as part of the breaking changes.

**Comment by wendorf** [comment]
@william-tran responses inline:

> On the presence guards, a template helper could be devised, thought needing to use dict to pass in params is also a bit ugly in its own way. Maybe list?

The Concourse team is working on [Investigate : Auto-magically generating K8s helm chart parameters](https://github.com/concourse/concourse/issues/2472) so I'm kinda tempted to punt on a perfect solution and just do ugly presence guards so the chart is more functional.

> [...] the ideal scenario is where secrets is kept to a minimum; they only contain credentials, so that the management overhead of these values is kept to a minimum [...] I can also see that this forces the chart maintainers to think about what actually is a secret and what isn't, but this is an important thing for us to make that determination for the benefit of chart users.

I'm on board, and can move non secrets out of this. I'll do my best determination of what is a secret and not, and hopefully get feedback if there are any mistakes.

> In terms of moving from secrets.oidcAuth.clientsecret and oidcAuth.secrets.clientsecret, this is at best "6 of 1, half a dozen of the other", which doesn't warrant the change in structure.

I'm not sure if you're saying you don't care about the structure, or if you'd prefer to keep it as `secrets.oidcAuth.clientsecret`. If that's what you'd prefer, I'm on board. If you don't care, I'd personally like to move to `oidcAuth.secrets.clientsecret`.

I'm on vacation through the end of the month and won't have a lot of time to work on this PR, but will get to it ASAP.

**Comment by antonu17** [comment]
> @antonu17 `web.env` was renamed to `web.extraEnv` as part of the breaking changes.

@wendorf I see it in the deployment template. Should be also updated in values.yaml. It is currently commented out as `env:`

**Comment by xtremerui** [comment]
the concourse team @pivotal-jamie-klassen and me have been working on [this fork](https://github.com/concourse/charts) where there is a lot of duplicate effort with this PR. We tried to represent all possible configuration values in the helm chart, and we have a [CI Job](https://ci.concourse-ci.org/teams/main/pipelines/main/jobs/check-helm-chart-params) that is continuously checking that the chart and the CLI flags are always in sync. This inspired us to restructure the `values.yaml` in a number of backwards-incompatible ways in order to keep the structure more similar to the actual names of the CLI flags. @william-tran @wendorf could we get your thoughts? We are currently trying to set up a scenario in which we can continuously run testflight (the official integration test suite) against a helm deployment (it's still a little [finicky](https://ci.concourse-ci.org/teams/main/pipelines/main/jobs/k8s-testflight) right now).

**Comment by william-tran** [comment]
@xtremerui  @pivotal-jamie-klassen Oh wow I didn't know the Concourse team was so far along in the effort to officially maintain the chart. Looking at that fork, the changes are in line with what I wanted to see: removing command line args in favour of only using env vars, and keeping secrets separate. There's a bit of external postgres stuff I need to verify but it looks mostly all there. From the perspective of the future of this chart, having the Concourse team officially maintain the chart and include its testing and maintenance in their release process is the best possible outcome. 

For those reasons I'm in favour of working with the fork submitted by the Concourse team, as a means to get them onto the official maintainers list; though that doesn't preclude @wendorf or any others that make significant contributions from being official maintainers as well. I do feel bad that the timing of your work and that of the Concourse team has resulted in this duplication of effort. But all the more reason to enjoy your vacation and not worry about getting this PR done.

**Comment by AndrewMorsillo** [comment]
@antonu17 I had indeed forgotten to add externalUrl to my yaml. I added it and did a helm upgrade but the deployment does not seem to reflect the changes. Will I need to destroy and re-create my deployment to change that value?

**Comment by william-tran** [comment]
Now that the concourse team PR #7804 has passed CI, I'm going to close this one, could all that have tested this PR try out that one? Again, this is the best move for the future maintenance of this chart. @ralekseenkov @AndrewMorsillo @antonu17 @chrishiestand @rubenv @arbourd @kurlzor @pms1969

**Comment by william-tran** [comment]
Superseded by #7804

**Comment by wendorf** [comment]
👍 It's never fun to see a big PR like this get closed without merging, but really excited to see more official support for Concourse on Kubernetes! Definitely the right thing!

**Pull request closed without merging**
